src/Voron/Util/ThreadHoppingReaderWriterLock.cs

```diff
@@ -99,13 +99,17 @@ private bool TryEnterReadLockSlow(int timeout)
                     ForTestingPurposes?.BeforeResetOfReaderWait?.Invoke();
 
                     if (_readerWait.IsSet)
+                    {
                         _readerWait.Reset();
 
-                    if (TryEnterReadLockCore())
-                    {
-                        // we got the reader lock, we can release all the other readers that might be waiting
-                        _readerWait.Set();
-                        return true;
+                        if (TryEnterReadLockCore())
```

Comment:
Now you are trying to call `TryEnterReadLockCore` only if this is set?
Isn't that now a race condition on `IsSet` ?

Reply:
If it isn't set, we wait on the `_readerWait`.
If it's set and we get the lock (`TryEnterReadLockCore`), we set it to release all the other readers.
If it isn't set, we wait on the `_readerWait`.

We should try to `TryEnterReadLockCore` only if it's set because we are here since we couldn't get the lock if the first place. So we are waiting for a signal (from the writer) in order to try to acquire the lock.